projects/bird-atlas/src/app/+grid/grid-index/grid-index-map/grid-index-map.component.ts

```diff
@@ -0,0 +1,184 @@
+  if (speciesCount < 150) { return scGradient[2]; }
+  return scGradient[3];
+};
+
+const gridSqToFeature = (square: AtlasGridSquare) => {
+  const latLngStr = square.coordinates.split(':');
+  return convertYkjToGeoJsonFeature(latLngStr[0], latLngStr[1]);
+};
+
+const getFeatureCollection = (grid: AtlasGrid) => ({
+  features: [
+    ...grid.map(square => <any>gridSqToFeature(square))
+  ],
+  type: 'FeatureCollection'
+});
+
+const getGetFeatureStyle = (grid: AtlasGrid, colorMode: ColorMode) => (
+  (opt) => ({
+    weight: 0,
+    opacity: 0,
+    fillOpacity: .5,
+    color: '#' + (
+      colorMode === 'activityCategory'
+        ? atlasActivityCategoryColors[grid[opt.featureIdx].activityCategory.key]
+        : getSpeciesCountColor(grid[opt.featureIdx].speciesCount)
+    )
+  })
+);
+
+const speciesCountLegendLabels = [
+  '1-49 lajia',
+  '50-99 lajia',
+  '100-149 lajia',
+  '150+ lajia',
+];
+
+const activityCategoryLegendLabels = {
+  'MY.atlasActivityCategoryEnum0': 'Ei havaintoja',
+  'MY.atlasActivityCategoryEnum1': 'Satunnaishavaintoja',
+  'MY.atlasActivityCategoryEnum2': 'Välttävä',
+  'MY.atlasActivityCategoryEnum3': 'Tyydyttävä',
+  'MY.atlasActivityCategoryEnum4': 'Hyvä',
+  'MY.atlasActivityCategoryEnum5': 'Erinomainen'
+};
+
+const legends: Record<ColorMode, { color: string; label: string }[]> = {
+  speciesCount: Object.entries(scGradient).map(([key, val]) => ({
+    color: val,
+    label: speciesCountLegendLabels[key]
+  })),
+  activityCategory: Object.entries(atlasActivityCategoryColors).map(([key, val]) => ({
+    color: val,
+    label: activityCategoryLegendLabels[key]
+  }))
+};
+
+@Component({
+  selector: 'ba-grid-index-map',
+  templateUrl: './grid-index-map.component.html',
+  styleUrls: ['./grid-index-map.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class GridIndexMapComponent implements AfterViewInit, OnDestroy {
+  private unsubscribe$ = new Subject<void>();
+
+  @Input() set atlasGrid(grid: AtlasGrid) {
+    if (environment.production === false) {
+      // populate test data
+      grid.forEach(g => {
+        g.speciesCount = Math.random() * 200;
+        const ac = Object.entries(activityCategoryLegendLabels)[Math.round(Math.random() * 5)];
+        g.activityCategory = {key: ac[0], value: ac[1]};
+      });
+    }
+    this.mapData$.next(this.getMapData(grid));
+  }
+  @Output() selectYKJ = new EventEmitter<string>();
+
+  @ViewChild('lajiMap', { static: false }) lajiMapElem: ElementRef;
+  private _colorMode: ColorMode = 'activityCategory';
+  set colorMode(m: ColorMode) {
+    this._colorMode = m;
+    const d = this.mapData$.getValue(); // mutate previous mapData for performance reasons
+    d.data.getFeatureStyle = getGetFeatureStyle(d.grid, this.colorMode);
+    this.mapData$.next(d);
+  }
+  get colorMode() {
+    return this._colorMode;
+  }
+
+  legends = legends;
+
+  private mapData$ = new BehaviorSubject<MapData>(undefined);
+  private map: any;
```

Comment:
This could look a little bit cleared if there was a clear order (like all private variables are next to each other). Also, we usually place the functions last or after the constructor as it makes easier to see the variables at a glance if the functions and variables are not mixed.

Reply:
I agree, but the issue with moving the atlasGrid setter is that then the inputs (and outputs) of the component are not at the same place... I made some changes:
* both map and table: removed setter from `atlasGrid` and moved the setter logic to `ngOnChanges`
* both map and table : reorganized the class variables to the following order:
  * `@Input`
  * `@Output`
  * template variables etc.
  * public variables
  * private variables
* table: moved cols intialization to the constructor
* table: proper typing for datatable rows
* map: moved colorMode setter logic to a private method so that the setter can be one lined
* map: refactored activity category colors so that the logic is equivalent to species count colors
* map: inlined precomputed gradients as was mentioned in original PR description. There's no reason to re-compute the gradients in runtime. Btw I do understand that writing the gradient logic was unnecessary and we could've just used some third party tool to pre-compute the gradients, but it was fun so I wanted to do it 😄

Reply:
I think the component looks very good and clear now :smiley: It doesn't matter so much where the functions are if they are one-lined/short